frontend/javascripts/oxalis/model/bucket_data_handling/pushqueue.ts

```diff
@@ -1,74 +1,91 @@
 import _ from "lodash";
 import type { DataBucket } from "oxalis/model/bucket_data_handling/bucket";
-import { alert, document } from "libs/window";
-import { sendToStore } from "oxalis/model/bucket_data_handling/wkstore_adapter";
-import AsyncTaskQueue from "libs/async_task_queue";
+import { createCompressedUpdateBucketActions } from "oxalis/model/bucket_data_handling/wkstore_adapter";
 import type DataCube from "oxalis/model/bucket_data_handling/data_cube";
-import Toast from "libs/toast";
-export const COMPRESSING_BATCH_SIZE = 32;
-// Only process the PushQueue after there was no user interaction
-// for PUSH_DEBOUNCE_TIME milliseconds...
+import { createDebouncedAbortableParameterlessCallable } from "libs/async/debounced_abortable_saga";
+import { call } from "redux-saga/effects";
+import Store from "oxalis/store";
+import { pushSaveQueueTransaction } from "../actions/save_actions";
+import { UpdateAction } from "../sagas/update_actions";
+import { AsyncFifoResolver } from "libs/async/async_fifo_resolver";
+import { escalateErrorAction } from "../actions/actions";
+
+// Only process the PushQueue after there was no user interaction (or bucket modification due to
+// downsampling) for PUSH_DEBOUNCE_TIME milliseconds.
+// PushQueue.getTransactionWaitTime is used to avoid debounce-related starvation that can happen
+// when the user constantly annotates.
 const PUSH_DEBOUNCE_TIME = 1000;
-// ...unless a timeout of PUSH_DEBOUNCE_MAX_WAIT_TIME milliseconds
-// is exceeded. Then, initiate a push.
-const PUSH_DEBOUNCE_MAX_WAIT_TIME = 30000;
 
 class PushQueue {
-  // @ts-expect-error ts-migrate(2564) FIXME: Property 'dataSetName' has no initializer and is n... Remove this comment to see the full error message
-  dataSetName: string;
   cube: DataCube;
-  compressionTaskQueue: AsyncTaskQueue;
-  sendData: boolean;
-  // The pendingQueue contains all buckets which are marked as
-  // "should be snapshotted and saved". That queue is processed
-  // in a debounced manner and sent to the `compressionTaskQueue`.
-  // The `compressionTaskQueue` compresses the bucket data and
-  // sends it to the save queue.
-  pendingQueue: Set<DataBucket>;
-
-  constructor(cube: DataCube, sendData: boolean = true) {
+
+  // The pendingQueue contains all buckets that should be:
+  // - snapshotted,
+  // - put into one transaction and then
+  // - saved
+  // That queue is flushed in a debounced manner so that the time of the
+  // snapshot should be suitable for a transaction (since neither WK nor the
+  // user edited the buckets in a certain time window).
+  private pendingQueue: Set<DataBucket>;
```

Comment:
Why is the queue a `Set`?

Reply:
The order of the buckets doesn't matter within a transaction, which is why this is a set. However, the mismatch is obviously still an eye sore. I renamed it to `pendingBuckets` now. The class `PushQueue` might also be ripe for a new name, but then the `PullQueue` loses its "name-sibling" :sweat_smile: Also, I'm not sure how to call it. `BucketTransactionBundler`? Might be okay to keep the PushQueue as is.